pkg/kube/informerfactory/factory.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package informerfactory provides a "factory" to generate informers. This allows users to create the
+// same informers in multiple different locations, while still using the same underlying resources.
+// Additionally, aggregate operations like Start, Shutdown, and Wait are available.
+// Kubernetes core has informer factories with very similar logic. However, this has a few problems that
+// spurred a fork:
+// * Factories are per package. That means we have ~6 distinct factories, which makes management a hassle.
+// * Across these, the factories are often inconsistent in functionality. Changes to these takes >4 months.
+// * Lack of functionality we want (see below).
+//
+// Added functionality:
+// * Single factory for any type, including dynamic informers, meta informers, typed informers, etc.
+// * Ability to create multiple informers of the same type but with different filters.
+// * Ability to run a single informer rather than all of them.
+package informerfactory
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// NewInformerFunc returns a SharedIndexInformer.
+type NewInformerFunc func() cache.SharedIndexInformer
+
+type StartableInformer struct {
+	Informer cache.SharedIndexInformer
+	start    func(stopCh <-chan struct{})
+}
+
+func (s StartableInformer) Start(stopCh <-chan struct{}) {
+	s.start(stopCh)
+}
+
+// InformerFactory provides access to a shared informer factory
+type InformerFactory interface {
+	// Start initializes all requested informers. They are handled in goroutines
+	// which run until the stop channel gets closed.
+	Start(stopCh <-chan struct{})
+
+	// InformerFor returns the SharedIndexInformer the provided type.
+	InformerFor(resource schema.GroupVersionResource, opts kubetypes.InformerOptions, newFunc NewInformerFunc) StartableInformer
+
+	// WaitForCacheSync blocks until all started informers' caches were synced
+	// or the stop channel gets closed.
+	WaitForCacheSync(stopCh <-chan struct{}) bool
```

Comment:
What for? I think we donot have a controller need to wait all cache sync

Reply:
This is what is called by `RunAndWait`